### DISCUSSION TITLE
Multi display: ST7789 + SSD1306

### DISCUSSION THREAD

**Discussion by valdus2** [arendst/Tasmota#21055] [Q&A] [answered]
I use an ST77889 with Universal driver and it works fine.
If I add an I2C oled SSD1306, it sees it in I2Cscan as a device with address 0x3c, but it doesn't use it as a display.

As per **gemu2015**: 
@Jason2866 
> Correct, Tasmota does not support more than one display.

to clarify this:

LVGL and berry support only one display

displaytext supports up to 4 simultaneous displays
with the following requirement:
all drivers except the first one must be renderer based drivers.

so you may combine e.g. one TM1637 (none renderer based)
with one SD1306 (renderer based)

_Originally posted by @gemu2015 in https://github.com/arendst/Tasmota/discussions/16950#discussioncomment-4143034_
 

Because I don't know what is a renderer based driver, can they work these two together with displaytext, and if so, how?

Thank you.

**Comment by gemu2015** [comment]
these are both renderer displays and you may combine them. there is an another thread how to do that
#13161

to further clarify:
you may also use multiple displays with berry and  LVGL.
but in this case the second display would be display text only and must be a renderer display

**Comment by valdus2** [comment]
I've seen this thread but is talking about SPI multidisplay.
I've seen also the other thread [#13222](13222) which is closer to my situation but clearly states that:

_In slot 1, you can connect any Tasmota supported display (non-universal display). You can connect any universal driver to slot 2 and slot 3._

Not only that, but it also uses the inbuilt driver for epaper. If I use the inbuilt driver for st7789 only garbage I see.

Now my problem seems way more complicated and i'm really confused what do I have to do.

Things I tried:

Everything that thread [#13222](13222) talks about. First I connected the I2C oled and then tried to load st7789 with `rule1 on System#Boot do DisplayText[S2/ST7789_display.ini:]`
or
`rule1 on Power1#Boot do DisplayText[S2/ST7789_display.ini:]`

Either software SPI or hardware SPI. Nothing at all.

Clean boot with the 2 displays connected with display.ini and optionA 3 for st7789 and simply connecting the I2C oled in I2C bus.After restarting this is weblog 4:

_00:00:00.003 HDW: ESP32-D0WD-V3 v3.0 
00:00:00.026 UFS: FlashFS mounted with 252 kB free
00:00:00.037 CFG: Loaded from File, Count 39
00:00:00.047 QPC: Count 1
00:00:00.048 CFG: CR 470/699, Busy 0
00:00:00.060 SPI: Hardware using GPIO18(CLK) and GPIO23(MOSI)
00:00:00.061 I2C: Bus1 using GPIO22(SCL) and GPIO21(SDA)
00:00:00.064 CFG: CR 470/699, Busy 0
00:00:00.065 CFG: CR 470/699, Busy 0
00:00:00.069 TFS: File 'mcp23x.dat' not found
00:00:00.070 MCP: Invalid template
00:00:00.070 TYA: Active=0
00:00:00.077 ROT: Mode 1
00:00:00.089 DSP: File descriptor used
00:00:01.068 DSP: Configured display 'ST7789'
00:00:01.074 MP3: Serial UART2
00:00:08.076 SRC: Restart
00:00:08.088 Project tasmota - Tasmota Version 13.4.0.3(DigiESP32v24_2)-2_0_14(2024-03-27T23:02:35)
00:00:08.090 Init audio I2S: port=0, bclk=27, ws=26, dout=25, mclk=-1, din=-1
00:00:08.104 ETH: No ETH MDC and ETH MDIO GPIO defined
00:00:08.116 SHT: Sensor did not ACK command
00:00:08.164 I2C: SHTC3 found at 0x70
00:00:08.168 iBeacon register for advert callbacks
00:00:08.170 MI32Option0: 1
00:00:08.170 MI32Option1: 0
00:00:08.171 MI32Option2: 0
00:00:08.171 MI32Option4: 1
00:00:08.172 MI32Option5: 0
00:00:08.182 MI32Option6: 0
00:00:08.183 M32: init: request callbacks
00:00:08.467 WIF: Checking connection...
00:00:08.468 WIF: Attempting connection...
00:00:10.000 WIF: Connecting to AP1 ASUS Channel 5 BSSId 40:16:7E:62:57:C0 in mode 11n as tasmota-DE6774-1908...
00:00:11.702 WIF: Checking connection...
00:00:11.703 WIF: Attempting connection...
00:00:12.714 WIF: Checking connection...
00:00:12.715 WIF: Attempting connection...
00:00:13.725 WIF: Checking connection...
00:00:13.726 WIF: Attempting connection...
00:00:14.736 WIF: Checking connection...
00:00:14.737 WIF: Attempting connection...
00:00:15.250 QPC: Reset
00:00:15.695 WIF: Checking connection...
00:00:15.696 WIF: Attempting connection...
00:00:16.633 WIF: IPv4 192.168.0.61, mask 255.255.255.0, gateway 192.168.0.254
00:00:16.707 WIF: Checking connection...
00:00:16.707 WIF: Connected
00:00:16.966 HTP: Web server active on tasmota-DE6774-1908 with IP address 192.168.0.61
00:00:17.216 APP: Boot Count 9
00:00:17.217 NTP: Sync time...
00:00:17.269 WIF: DNS resolved '2.pool.ntp.org' (91.217.155.60) in 51 ms
00:00:17.357 RTC: UTC 2024-03-28T17:37:33Z, DST 2024-03-31T03:00:00, STD 2024-10-27T04:00:00
19:37:33.001 RTC: Synced by NTP
19:37:33.198 CFG: Saved, Count 40, Bytes 4096
19:37:33.510 HTP: Main Menu
19:37:34.003 MQT: Attempting connection...
19:37:34.005 WIF: DNS resolved '192.168.0.5' (192.168.0.5) in 0 ms
19:37:34.029 MQT: Connected
19:37:34.034 MQT: tele/tasmota_DE6774/LWT = Online (retained)
19:37:34.039 MQT: cmnd/tasmota_DE6774/POWER = 
19:37:34.041 MQT: Subscribe to cmnd/tasmota_DE6774/#
19:37:34.043 MQT: Subscribe to cmnd/tasmotas/#
19:37:34.044 MQT: Subscribe to cmnd/DVES_DE6774_fb/#
19:37:34.056 MQT: tele/tasmota_DE6774/INFO1 = {"Info1":{"Module":"ESP32-DevKit","Version":"13.4.0.3(DigiESP32v24_2)","FallbackTopic":"cmnd/DVES_DE6774_fb/","GroupTopic":"cmnd/tasmotas/"}}
19:37:34.077 MQT: tele/tasmota_DE6774/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-DE6774-1908","IPAddress":"192.168.0.61"}}
19:37:34.094 MQT: tele/tasmota_DE6774/INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":10}}
19:37:34.101 MQT: stat/tasmota_DE6774/RESULT = {"POWER1":"OFF"}
19:37:34.107 MQT: stat/tasmota_DE6774/POWER1 = OFF
19:37:34.109 MQT: stat/tasmota_DE6774/RESULT = {"POWER2":"ON"}
19:37:34.111 MQT: stat/tasmota_DE6774/POWER2 = ON
19:37:34.128 TFS: File 'autoexec.bat' not found
19:37:35.625 HTP: Tools
19:37:36.801 HTP: Console
19:37:38.210 MQT: tele/tasmota_DE6774/STATE = {"Time":"2024-03-28T19:37:38","Uptime":"0T00:00:15","UptimeSec":15,"Heap":54,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":21,"MqttCount":1,"POWER1":"OFF","POWER2":"ON","Wifi":{"AP":1,"SSId":"ASUS","BSSId":"40:16:7E:62:57:C0","Channel":5,"Mode":"11n","RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:09"}}
19:37:38.314 MQT: tele/tasmota_DE6774/SENSOR = {"Time":"2024-03-28T19:37:38","Switch1":"ON","ANALOG":{"Range1":95,"MQ135_2":0.04},"SHTC3":{"Temperature":20.16,"Humidity":65.5,"DewPoint":13.47,"HeatIndex":19.94},"ESP32":{"Temperature":53.33},"TempUnit":"C"}
19:37:38.350 MQT: tele/tasmota_DE6774/BLE = {"Time":"2024-03-28T19:37:38","BLEDevices":{"total":0}}
19:37:38.357 MQT: tele/tasmota_DE6774/BLE = {"Time":"2024-03-28T19:37:38","BLE":{"scans":0,"adverts":0,"devices":0,"resets":0}}
19:37:43.236 MQT: tasmota/discovery/B0A732DE6774/config = {"ip":"192.168.0.61","dn":"Tasmota","fn":["Tasmota","Tasmota2",null,null,null,null,null,null],"hn":"tasmota-DE6774-1908","mac":"B0A732DE6774","md":"ESP32-DevKit","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"13.4.0.3","t":"tasmota_DE6774","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"bat":0,"dslp":0,"sho":[],"sht":[],"ver":1} (retained)
19:37:43.313 MQT: tasmota/discovery/B0A732DE6774/sensors = {"sn":{"Time":"2024-03-28T19:37:43","Switch1":"ON","ANALOG":{"Range1":95,"MQ135_2":0.04},"SHTC3":{"Temperature":20.14,"Humidity":65.4,"DewPoint":13.44,"HeatIndex":19.92},"ESP32":{"Temperature":33.33},"TempUnit":"C"},"ver":1} (retained)_


No sign from I2c display. If I run I2cscan, it finds it fine in address 0x3c:

_19:40:28.410 CMD: i2cscan
19:40:28.412 SRC: WebConsole from 192.168.0.102
19:40:28.415 CMD: Grp 0, Cmd 'I2CSCAN', Idx 1, Len 0, Pld -99, Data ''
19:40:28.443 MQT: stat/tasmota_DE6774/RESULT = {"I2CScan":"Device(s) found on bus1 at 0x3c 0x70"}_


This is my config: (the relay is a physical external relay)
![11](https://github.com/arendst/Tasmota/assets/150083707/af39d3da-2d98-4f06-9b15-9422beba8724)


And this is what I get:, only the switch1 for the physical relay and switch2 for st7789:
![22](https://github.com/arendst/Tasmota/assets/150083707/b7a9bfd3-6c10-4bd3-8939-24534cb1c894)



If someone could help me with really simple steps, I'd really appreciated it, cause I'm totally lost....
Thank you.

**Comment by valdus2** [comment]
A new attempt, and this is how far I can go:

_00:00:00.003-163/35 HDW: ESP32-D0WD-V3 v3.0 
00:00:00.135-162/34 UFS: FlashFS mounted with 252 kB free
00:00:00.165 CFG: Loaded from File, Count 81
00:00:00.176 QPC: Count 1
00:00:00.177 CFG: CR 470/699, Busy 0
00:00:00.189 SPI: Software using GPIO18(CLK) and GPIO23(MOSI)
00:00:00.190 I2C: Bus1 using GPIO22(SCL) and GPIO21(SDA)
00:00:00.193 CFG: CR 470/699, Busy 0
00:00:00.194 CFG: CR 470/699, Busy 0
00:00:00.201 TFS: File 'mcp23x.dat' not found
00:00:00.201 MCP: Invalid template
00:00:00.202 TYA: Active=0
00:00:00.209 ROT: Mode 1
**00:00:00.210 I2C: SSD1306 found at 0x3c
00:00:00.251 DSP: SSD1306
00:00:00.308 DSP: File descriptor used
00:00:00.309 DSP: reinit
00:00:01.255 DSP: Configured display 'ST7789'**
00:00:01.263 MP3: Serial UART2
00:00:08.265 SRC: Restart
00:00:08.278 Project tasmota - Tasmota Version 13.4.0.3(DigiESP32v24_2)-2_0_14(2024-03-27T23:02:35)_

The ST7789 is loaded with `rule1 on Power1#Boot do DisplayText[S2/ST7789_display.ini:`

After that only st7789 works. Stumbled...

**Comment by gemu2015** [comment]
did you enable USE_MULTI_DISPLAY ?
you should get a special log message on DisplayText[S2/ST7789_display.ini:]
DSP: File: ST7789_display.ini
and DSP: File descriptor loaded

do not do a display reinit.

**Reply by valdus2** [reply]
Thanks. Yes of course I have enabled multi display, a part of my user_config_override.h below.

I started clean with only ssd1306 in I2C bus. It worked OK with displaymodel 2.
Then I enabled gpio0 option a3 for universal driver (for st7789) and I added a rule to load the ST7789_display.ini:
`rule1 on Power1#Boot do DisplayText[S2/ST7789_display.ini:]`
After that only ST7789 works. Either command DisplayText[S1:] or DisplayText[S2:] sends to ST7789.
And there is no button in webUI to toggle the SSD1306.

I don't do manually a reinit (I don't know how), tasmota did it after loading the ST7789_display.ini and then the  ssd1306 vanishes as a display. But in I2Cscan exists.

And this is weblog 4 after a restart.



```
00:00:00.003 HDW: ESP32-D0WD-V3 v3.0 
00:00:00.086 UFS: FlashFS mounted with 248 kB free
00:00:00.107 CFG: Loaded from File, Count 116
00:00:00.117 QPC: Count 1
00:00:00.118 CFG: CR 470/699, Busy 0
00:00:00.130 SPI: Software using GPIO18(CLK) and GPIO23(MOSI)
00:00:00.132 I2C: Bus1 using GPIO22(SCL) and GPIO21(SDA)
00:00:00.134 CFG: CR 470/699, Busy 0
00:00:00.135 CFG: CR 470/699, Busy 0
00:00:00.141 TFS: File 'mcp23x.dat' not found
00:00:00.141 MCP: Invalid template
00:00:00.142 TYA: Active=0
00:00:00.148 ROT: Mode 1
00:00:00.182 DSP: File descriptor used
00:00:01.188 DSP: Configured display 'ST7789'
00:00:01.195 MP3: Serial UART2
00:00:08.197 SRC: Restart
00:00:08.210 Project tasmota - Tasmota Version 13.4.0.3(DigiESP32v24_2gr)-2_0_14(2024-03-29T00:03:40)
00:00:08.212 Init audio I2S: port=0, bclk=27, ws=26, dout=25, mclk=-1, din=-1
00:00:08.225 ETH: No ETH MDC and ETH MDIO GPIO defined
00:00:08.237 SHT: Sensor did not ACK command
00:00:08.285 I2C: SHTC3 found at 0x70
00:00:08.289 iBeacon register for advert callbacks
00:00:08.291 MI32Option0: 1
00:00:08.291 MI32Option1: 0
00:00:08.292 MI32Option2: 0
00:00:08.292 MI32Option4: 1
00:00:08.293 MI32Option5: 0
00:00:08.303 MI32Option6: 0
00:00:08.304 M32: init: request callbacks
00:00:08.588 WIF: Checking connection...
00:00:08.589 WIF: Attempting connection...
00:00:09.225 WIF: Connecting to AP1 ASUS Channel 5 BSSId 40:16:7E:62:57:C0 in mode 11n as tasmota-DE6774-1908...
00:00:14.856 WIF: Checking connection...
00:00:14.857 WIF: Attempting connection...
00:00:15.370 QPC: Reset
00:00:15.817 WIF: Checking connection...
00:00:15.817 WIF: Attempting connection...
00:00:16.697 WIF: IPv4 192.168.0.59, mask 255.255.255.0, gateway 192.168.0.254
00:00:16.828 WIF: Checking connection...
00:00:16.829 WIF: Connected
00:00:17.090 HTP: Web server active on tasmota-DE6774-1908 with IP address 192.168.0.59
00:00:17.136 HTP: Main Menu
00:00:17.402 APP: Boot Count 24
00:00:17.403 NTP: Sync time...
00:00:17.457 WIF: DNS resolved '2.pool.ntp.org' (162.159.200.1) in 53 ms
00:00:17.484 RTC: UTC 2024-03-29T15:36:42Z, DST 2024-03-31T03:00:00, STD 2024-10-27T04:00:00
17:36:42.000 RTC: Synced by NTP
17:36:42.251 CFG: Saved, Count 117, Bytes 4096
17:36:43.097 MQT: Attempting connection...
17:36:43.098 WIF: DNS resolved '192.168.0.5' (192.168.0.5) in 0 ms
17:36:43.159 MQT: Connected
17:36:43.165 MQT: tele/tasmota_DE6774/LWT = Online (retained)
17:36:43.168 MQT: cmnd/tasmota_DE6774/POWER = 
17:36:43.169 MQT: Subscribe to cmnd/tasmota_DE6774/#
17:36:43.171 MQT: Subscribe to cmnd/tasmotas/#
17:36:43.172 MQT: Subscribe to cmnd/DVES_DE6774_fb/#
17:36:43.181 MQT: tele/tasmota_DE6774/INFO1 = {"Info1":{"Module":"ESP32-DevKit","Version":"13.4.0.3(DigiESP32v24_2gr)","FallbackTopic":"cmnd/DVES_DE6774_fb/","GroupTopic":"cmnd/tasmotas/"}}
17:36:43.202 MQT: tele/tasmota_DE6774/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-DE6774-1908","IPAddress":"192.168.0.59"}}
17:36:43.221 MQT: tele/tasmota_DE6774/INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":25}}
17:36:43.228 MQT: stat/tasmota_DE6774/RESULT = {"POWER1":"OFF"}
17:36:43.233 MQT: stat/tasmota_DE6774/POWER1 = OFF
17:36:43.235 MQT: stat/tasmota_DE6774/RESULT = {"POWER2":"ON"}
17:36:43.248 MQT: stat/tasmota_DE6774/POWER2 = ON
17:36:43.255 TFS: File 'autoexec.bat' not found
17:36:44.234 HTP: Tools
17:36:45.523 HTP: Console
```






```
#ifndef USE_I2C
  #define USE_I2C
#endif

#ifndef USE_SPI
  #define USE_SPI
#endif

#ifndef USE_DISPLAY
  #define USE_DISPLAY
#endif

#ifndef USE_UNIVERSAL_DISPLAY
  #define USE_UNIVERSAL_DISPLAY
#endif

#ifndef USE_DISPLAY_MODES1TO5
  #define USE_DISPLAY_MODES1TO5
#endif

#ifndef USE_DISPLAY_LCD
  #define USE_DISPLAY_LCD
#endif

#ifndef USE_DISPLAY_SSD1306
  #define USE_DISPLAY_SSD1306
#endif

#ifndef USE_DISPLAY_ST7789
  #define USE_DISPLAY_ST7789
#endif

#ifndef USE_TOUCH_BUTTONS
  #define USE_TOUCH_BUTTONS
#endif

#ifndef USE_RAMFONT
  #define USE_RAMFONT
#endif

#ifndef USE_MULTI_DISPLAY
  #define USE_MULTI_DISPLAY
#endif

#ifndef USE_GRAPH
  #define USE_GRAPH
#endif

#ifndef NUM_GRAPHS
  #define NUM_GRAPHS 4
#endif

#ifndef USE_DT_VARS
  #define USE_DT_VARS
#endif

#ifndef SHOW_SPLASH
  #define SHOW_SPLASH
#endif

#ifndef USE_AWATCH
  #define USE_AWATCH
#endif

```

**Reply by gemu2015** [reply]
as already set, you do not get the right logging messages.
i would suggest you switch to scripting and use the example i provided in the other thread,
and since there is also a universal ini for ssd1306 i wouldn't use the native driver.

**Reply by valdus2** [reply]
I haven't used scripting, but I don't see any difference between the commands either way.
Now I thought so to use udriver for ssd1306 and here comes the reinit again.
I first load the ssd1306 and then when it loads st7789 comes the reinit and only st7789 works.


```
18:05:31.862 CMD: DisplayText[S1/SSD1306_display.ini:]
18:05:31.864 SRC: WebConsole from 192.168.0.102
18:05:31.865 CMD: Grp 0, Cmd 'DISPLAYTEXT', Idx 1, Len 25, Pld -99, Data '[S1/SSD1306_display.ini:]'
18:05:31.868 DSP: File: /SSD1306_display.ini
18:05:31.910 DSP: const char descriptor used
18:05:31.913 I2C: DSP-I2C found at 0x3c
18:05:32.047 DSP: Configured display 'SSD1306'
18:05:32.048 DSP: File descriptor loaded
18:05:32.178 MQT: stat/tasmota_DE6774/RESULT = {"DisplayText":"[S1/SSD1306_display.ini"}
18:05:32.793 CFG: Saved, Count 133, Bytes 4096
18:05:34.933 WIF: Checking connection...
18:06:31.207 M32: Kick off readOneSensor
18:06:31.209 M32: Kick off tele sending
18:06:31.961 CMD: DisplayText[S2/ST7789_display.ini:]
18:06:31.964 SRC: WebConsole from 192.168.0.102
18:06:31.966 CMD: Grp 0, Cmd 'DISPLAYTEXT', Idx 1, Len 24, Pld -99, Data '[S2/ST7789_display.ini:]'
18:06:31.970 DSP: File: /ST7789_display.ini
18:06:32.026 DSP: const char descriptor used
18:06:32.028 DSP: reinit
18:06:33.165 DSP: Configured display 'ST7789'
18:06:33.167 DSP: File descriptor loaded
18:06:33.175 MQT: stat/tasmota_DE6774/RESULT = {"DisplayText":"[S2/ST7789_display.ini"}
18:06:34.418 CFG: Saved, Count 135, Bytes 4096
18:06:36.261 WIF: Checking connection...
```

**Reply by gemu2015** [reply]
ok, at least you now have the appropriate logging response for the S cmd. (File: /ST7789_display.ini)
now you should be able to switch back and forth between the displays.
switch to S1, write something, switch to S2, write something etc.

**Reply by valdus2** [reply]
I tried right now the other way loading ssd1306 after st7789 and now a reinit again and only ssd1306 works.

So after loading the second driver only the lasst one works.

```

18:20:16.277 WIF: Checking connection...
18:20:20.701 CMD: DisplayText[S1/SSD1306_display.ini:]
18:20:20.703 SRC: WebConsole from 192.168.0.102
18:20:20.705 CMD: Grp 0, Cmd 'DISPLAYTEXT', Idx 1, Len 25, Pld -99, Data '[S1/SSD1306_display.ini:]'
18:20:20.708 DSP: File: /SSD1306_display.ini
18:20:20.724 DSP: const char descriptor used
18:20:20.726 DSP: reinit
18:20:20.859 DSP: Configured display 'SSD1306'
18:20:20.861 DSP: File descriptor loaded
18:20:20.988 MQT: stat/tasmota_DE6774/RESULT = {"DisplayText":"[S1/SSD1306_display.ini"}
18:20:21.347 CFG: Saved, Count 136, Bytes 4096
18:20:32.213 M32: Kick off readOneSensor
18:20:32.215 M32: Kick off tele sending
18:20:36.522 WIF: Checking connection...
```

**Reply by valdus2** [reply]
Yes I always do a test and when I say it doesn't work, the commands S1 Or S2 send only to the one display that works.

**Reply by gemu2015** [reply]
ok, i have not tested that for a long time and i can confirm that it no longer works, even crashes sometimes. it used to work stable and without problems.
i have to check what causes the problems and fix them. will take a while.

**Reply by valdus2** [reply]
OK, at least I'm not crazy...And I can confirm also that now that ssd1306 works sending displaytext to st7789 that doesn't work crashes my system also. 
Thank you gemu2015 for confirming this and I'm waiting for a solution.

**Reply by gemu2015** [reply]
ok, replace this file and check again

[xdrv_13_display.ino.zip](https://github.com/arendst/Tasmota/files/14807560/xdrv_13_display.ino.zip)

**Comment by valdus2** [comment]
I had the most thorough tests I have done in my life...believe it or not...


Lets start:

**1) Tasmota compiled with the native driver for ssd1306 and the new file from you.**


_a) First try using native driver for ssd1306 and universal driver for st7789_

First command displaymodel 2 because this the only way to see it after a restart as an I2C display with the native driver.
Restarting (to be exactly crashing when changing this command with two displays connected).
Second command loading manually st7789_display.ini
No reinit and the 2 displays are there, but, and this is a big BUT, everything is so much messed up in every command that I gave up.

_b) Second try using universal driver for ssd1306 and universal driver for st7789_

First command displaymodel 17 because this the only way to not use the I2C display with the native driver.
Restarting (to be exactly crashing when changing this command with two displays connected).
Second command `rule1 on Power1#Boot do Backlog DisplayText[S1/SSD1306_display.ini:]; DisplayText[S2/ST7789_display.ini:]`
Weblog 4 answer: `DSP: No valid descriptor found`
Third command `rule1 on System#Boot do Backlog DisplayText[S1/SSD1306_display.ini:]; DisplayText[S2/ST7789_display.ini:]`
Weblog 4 answer: `DSP: No valid descriptor found`

I empty the rules and restart.
After restarting i give it manually the command from console:
`Backlog DisplayText[S1/SSD1306_display.ini:]; DisplayText[S2/ST7789_display.ini:]`
Weblog 4 answer: 
```
01:12:32.447 CMD: Backlog DisplayText[S1/SSD1306_display.ini:]; DisplayText[S2/ST7789_display.ini:]
01:12:32.451 SRC: WebConsole from 192.168.0.102
01:12:32.453 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 73, Pld -99, Data 'DisplayText[S1/SSD1306_display.ini:]; DisplayText[S2/ST7789_display.ini:]'
01:12:32.490 SRC: Backlog
01:12:32.493 CMD: Grp 0, Cmd 'DISPLAYTEXT', Idx 1, Len 25, Pld -99, Data '[S1/SSD1306_display.ini:]'
01:12:32.496 DSP: File: /SSD1306_display.ini
01:12:32.550 DSP: const char descriptor used
01:12:32.551 I2C: DSP-I2C found at 0x3c
01:12:32.685 DSP: Configured display 'SSD1306'
01:12:32.687 DSP: File descriptor loaded
01:12:32.815 MQT: stat/tasmota_DE6774/RESULT = {"DisplayText":"[S1/SSD1306_display.ini"}
01:12:32.825 SRC: Backlog
01:12:32.827 CMD: Grp 0, Cmd 'DISPLAYTEXT', Idx 1, Len 24, Pld -99, Data '[S2/ST7789_display.ini:]'
01:12:32.831 DSP: File: /ST7789_display.ini
01:12:32.886 DSP: const char descriptor used
01:12:34.013 DSP: Configured display 'ST7789'
01:12:34.015 DSP: File descriptor loaded
01:12:34.023 MQT: stat/tasmota_DE6774/RESULT = {"DisplayText":"[S2/ST7789_display.ini"}

```

WTF!!! They work both of them with the below observations:

They need these commands to initialize properly
DisplayText[S1:]
Displayinvert 0
Displayrotate 0

DisplayText[S2:]
Displayinvert 1
Displayrotate 1

And the important part:
Only 1 button in webUI which if I close it, it turns off the first display and it never comes back on again without a restart.



**2) Tasmota compiled without the native driver for ssd1306 and the new file from you.**

Everything the same as my second try above.



And now comes the tricky part for me. How do I load them automatically?
Using autoexec.bat solved the problem. Having a display.ini with the code for ssd1306
and a command in autoexec.bat `DisplayText[S2/ST7789_display.ini:]`

The problem is the 1 button in webUI which if I close it, it turns off the first display and it never comes back on again without a restart.
_The use of RAM is about 1-1.5KB when they are used together to display real time values from sensors_

Thanks again for your efforts.

**Comment by gemu2015** [comment]
do not compile the native drivers at all.
as initial display.ini, use the most used display, normally the larger one.
on boot load the second display, draw initial things on it and immediately switch back to screen 1
screen 1 now works normal, including the power button.
if you want to draw something to screen2  switch to screen 2, draw and immediately switch back to 1.
so that screen 1 is always in focus. 
power button works as usual but only switches screen 1
if you want to switch power of screen 2, switch to 2 and use display text cmd [o] and [O] and then back to 1

**Comment by valdus2** [comment]
OK, I loaded first the st7789 as first display and following your steps above I can power On and Off the individual screens as you said.

Also I forgot to tell that this combination works either with software SPI or hardware SPI (change the descriptor and the module config).

So, I keep the new ino until implemented in a new release ?

Thank you.

**Reply by gemu2015** [reply, marked as answer]
ok is fixed and merged into development

**Comment by valdus2** [comment]
Thank you [gemu2015](https://github.com/gemu2015). I think we can close it.
